File: sdk-android/library/src/main/kotlin/app/uxclone/sdk/Transport.kt
package app.uxclone.sdk

import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit

/** Thread-safe event queue with timed flush to the backend. */
internal class Transport(
    private val config:       UXConfig,
    private val getSessionId: () -> String,
    private val getAnonId:    () -> String,
) {
    private val queue:     LinkedBlockingQueue<UXEvent> = LinkedBlockingQueue()
    private var scheduler: ScheduledExecutorService?   = null
    private var retryCount = 0

    fun push(event: UXEvent) {
        queue.offer(event)
        if (queue.size >= config.maxBatchSize) flush()
    }

    fun startAutoFlush() {
        val s = ScheduledThreadPoolExecutor(1)
        scheduler = s
        s.scheduleWithFixedDelay({ flush() }, config.flushInterval, config.flushInterval, TimeUnit.MILLISECONDS)
    }

    fun stopAutoFlush() {
        scheduler?.shutdown()
        scheduler = null
    }

    fun flush() {
        if (queue.isEmpty()) return
        val batch = mutableListOf<UXEvent>()
        queue.drainTo(batch, config.maxBatchSize)
        if (batch.isNotEmpty()) sendBatch(batch)
    }

    fun flushSync() {
        val batch = mutableListOf<UXEvent>()
        queue.drainTo(batch)
        if (batch.isNotEmpty()) sendBatch(batch, retries = 0)
    }

    private fun sendBatch(events: List<UXEvent>, retries: Int = 3) {
        Thread {
            try {
                val payload = JSONObject().apply {
                    put("sessionId",   getSessionId())
                    put("anonymousId", getAnonId())
                    put("apiKey",      config.apiKey)
                    put("events",      JSONArray(events.map { it.toJson() }))
                }
                val url  = URL("${config.endpoint}/api/v1/ingest/batch")
                val conn = url.openConnection() as HttpURLConnection
                conn.requestMethod = "POST"
                conn.setRequestProperty("Content-Type", "application/json")
                conn.doOutput = true
                conn.outputStream.use { it.write(payload.toString().toByteArray()) }

                val code = conn.responseCode
                if (code >= 500 && retries > 0) {
                    Thread.sleep(1_000L * (4 - retries))
                    sendBatch(events, retries - 1)
                }
                conn.disconnect()
            } catch (e: Exception) {
                if (retries > 0) {
                    Thread.sleep(1_000L)
                    sendBatch(events, retries - 1)
                }
            }
        }.apply { isDaemon = true; start() }
    }
}

File: sdk-ios/Sources/UXClone/Transport.swift
import Foundation

/// Thread-safe HTTP transport with automatic batching and retry.
final class Transport {
    private var queue:        [UXEvent]        = []
    private let queueLock                      = NSLock()
    private var flushTimer:   Timer?
    private let config:       UXConfig
    private let getSessionId: () -> String
    private let getAnonId:    () -> String
    private let session:      URLSession

    init(config: UXConfig, getSessionId: @escaping () -> String, getAnonId: @escaping () -> String) {
        self.config       = config
        self.getSessionId = getSessionId
        self.getAnonId    = getAnonId

        // Use background session so requests survive app backgrounding
        let cfg = URLSessionConfiguration.background(withIdentifier: "app.uxclone.sdk")
        cfg.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: cfg)
    }

    func push(_ event: UXEvent) {
        queueLock.lock()
        queue.append(event)
        let count = queue.count
        queueLock.unlock()

        if count >= config.maxBatchSize { flush() }
    }

    func startAutoFlush() {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.flushTimer = Timer.scheduledTimer(withTimeInterval: self.config.flushInterval, repeats: true) { [weak self] _ in
                self?.flush()
            }
        }
    }

    func stopAutoFlush() {
        flushTimer?.invalidate()
        flushTimer = nil
    }

    func flush() {
        queueLock.lock()
        guard !queue.isEmpty else { queueLock.unlock(); return }
        let batch = Array(queue.prefix(config.maxBatchSize))
        queue.removeFirst(min(config.maxBatchSize, queue.count))
        queueLock.unlock()

        sendBatch(batch)
    }

    private func sendBatch(_ events: [UXEvent], retryCount: Int = 0) {
        guard let url = URL(string: "\(config.endpoint)/api/v1/ingest/batch") else { return }

        let payload: [String: Any] = [
            "sessionId":   getSessionId(),
            "anonymousId": getAnonId(),
            "apiKey":      config.apiKey,
            "events":      events.compactMap { try? JSONSerialization.jsonObject(with: JSONEncoder().encode($0)) },
        ]
        guard let body = try? JSONSerialization.data(withJSONObject: payload) else { return }

        var req = URLRequest(url: url)
        req.httpMethod  = "POST"
        req.httpBody    = body
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")

        session.dataTask(with: req) { [weak self] _, response, error in
            guard let self = self else { return }
            let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
            if error != nil || (statusCode >= 400 && statusCode < 500 == false) {
                if retryCount < 3 {
                    DispatchQueue.global().asyncAfter(deadline: .now() + Double(retryCount + 1)) {
                        self.sendBatch(events, retryCount: retryCount + 1)
                    }
                }
            }
        }.resume()
    }
}
